### PR TITLE
Improve the VFS extra metadata handling

### LINKED ISSUE #7620: Rename "metadata" to "localized" in the VFS file metadata
### Affected version(s)

4.13, 5.3

### Description

When using the VFS you can get "extra" metadata, which is an arbitrary collection of things read from a DBAFS and enriched in the `RetrieveMetadataEvent`:

```php
$metadata = $file->getExtraMetadata(); // array<string, mixed>
```

In there are things (the core populates), that are not localized and a collection of localized things: 
```php
// Not localized
$importantPart = $metadata['importantPart']; // ImportantPart
$subtitle = $metadata['subtitle']; // Subtitle

// Localized
$localizedMetadata = $metadata['metadata']; // MetadataBag
$localizedMetadata->get('fr')->getTitle(); // string
``` 

It is a bit confusing that there is metadata in extra metadata and that **that** is the localized bit. IMHO it would be much better DX-wise if the latter had the key `localized` instead of `metadata`:
```php
$localizedMetadata = $metadata['localized']; // MetadataBag
``` 

We could either just make this change (experimental feature) _or_ populate both keys for BC and remove `metadata` later on. One downside for having both is, that then there are also two places where you can change things leading to inconsistent data.

### PR DESCRIPTION
**Pull request by m-vo**
Fixes #7620

I'll provide a separate PR for 4.13 after this has been approved/merged.

### Conceptional changes
When dealing with "extra metadata", i.e. additional data coming from adapters or any DBAFS source, you would get an array with string keys and arbitrary values. A big downside is, that this disallows chained access like `$item->getExtraMetadata()->get('foo')?->getBar()?->getBaz()`. Even for core features you also had to know the keys and cannot use convenience getters.

This affects working with the following methods:
* `VirtualFilesystemInterface#getExtraMetadata()`
* `VirtualFilesystemInterface#setExtraMetadata()`
* `FilesystemItem#getExtraMetadata()`
* `DbafsInterface#setExtraMetadata()`
* `AbstractDbafsMetadataEvent#getExtraMetadata()`

Now instead of an array, we've got a class that implements `\ArrayAccess`: `ExtraMetadata`. 

I also added some helper methods for the things the core's `DbafsMetadataSubscriber` is adding (`tl_files`). For any other place building `ExtraMetadata` these just return `null` (like when there is no data). I thought about having a `FileExtraMetadata` or something that has the core's convenience features but as extra metadata can come from multiple sources _at once_ and needs to be merged, this would need a similar pattern with multiple bags like we have for the response context - way overkill here and the usage would be worse.

### Name change
The core `DbafsMetadataSubscriber` returned a key `metadata` under which the localized metadata was found in a `MetadataBag`. This was confusing: why "metadata" inside "extra metadata"? It's now renamed to `localized` + there is a convenience accessor `getLocalized()`. Setting or reading `metadata` still works but now triggers a deprecation.

### Impact
This is obviously a BC break affecting custom implementations _but_ anyone just using the results should be fine. As the VFS is an experimental feature, I'd strongly vote for breaking things now before having to deal with the inconsistencies later on.

### CODE REVIEW

**Comment by m-vo** [comment]
Rector is currently failing, but its code suggestion would generate a runtime error. The issue could be circumvented by adding a `(bool)` cast but this would be semantically wrong in my opinion.

File: core-bundle/src/Controller/ContentElement/DownloadsController.php [hunk 1]
```diff
@@ -96,8 +96,7 @@ protected function getFilesystemItems(Request $request, ContentModel $model): Fi
         // Optionally filter out files without metadata
         if ('downloads' === $model->type && $model->metaIgnore) {
             $filesystemItems = $filesystemItems->filter(
-                static fn (FilesystemItem $item): bool => null !== ($metadata = $item->getExtraMetadata()['metadata'] ?? null)
-                    && null !== $metadata->getDefault(),
+                static fn (FilesystemItem $item): bool => null !== $item->getExtraMetadata()->getLocalized()?->getDefault(),
```
Comment (m-vo) [inline core-bundle/src/Controller/ContentElement/DownloadsController.php]:
```suggestion
                static fn (FilesystemItem $item): bool => (bool) $item->getExtraMetadata()->getLocalized()?->getDefault(),
```

**Comment by m-vo** [comment]
@Toflar Could you check if everything in the notification center is still working with these changes? IMHO it should from reading the code, but this would be a good indicator that the BC break is indeed not that big of a deal.

**Review by Toflar — COMMENTED** [review]
Looks good to me as well. Even if it would cause issues, the VFS is experimental so no worries :)

**Review by Toflar — DISMISSED** [review]
I mean, I will have to adjust the tests because I've implemented the `DbafsInterface` for tests but that's fine.

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @m-vo.

**Pull request merged**